### DISCUSSION TITLE
ISPC grammar for tree-sitter language parser

### DISCUSSION THREAD

**Discussion by fab4100** [ispc/ispc#2527] [Show and tell]
Dear All,

I have written some grammar for an ISPC language parser in [tree-sitter](https://tree-sitter.github.io/tree-sitter/).  The grammar and parser source can be found in [https://github.com/fab4100/tree-sitter-ispc](https://github.com/fab4100/tree-sitter-ispc).

The parser and various queries are now also integrated in [nvim-treesitter](https://github.com/nvim-treesitter/nvim-treesitter/commit/8e817bebf9184cab6c59205e1aed3fdce11c8c51).

**Comment by dbabokin** [comment]
Thanks for this contribution!

Looking at the repo, I see that you have some tests, but they are not fully covering the language. It's probably ok, but i would suggest specifically covering floating point literals - we parse them differently as in C/C++. You can use these to tests for the examples of the literals that need to be parsed: [float_literals.ispc](https://github.com/ispc/ispc/blob/main/tests/lit-tests/float_literals.ispc) and [float_literals_value.ispc](https://github.com/ispc/ispc/blob/main/tests/lit-tests/float_literals_value.ispc)

Also, I'm wondering what is the process for updating the grammar? I think we are not going to add anything new in the short term, but would be good to understand the process.

**Reply by fab4100** [reply]
I have only added tests for the grammar rules that were needed on top of what is implemented for C already, but you are correct that tests for number literals are missing.  I will add these tests since I did implement a new rule for number literals in ISPC.  The current grammar parses everything in `stdlib.ispc` and all code below `tests/` without errors in the syntax trees.  There are a few exceptions in `tests/lit-tests`: there are some syntax tree errors that are excepted (e.g. in `float_literals_errors.ispc`) and some are due to interplay between pre-processor constructs and ISPC code (which would result in correct IPSC code after the `cpp` pass).  One other issue that currently results in syntax tree error is `foreach (int i = 0...N)` versus `foreach (int i = 0 ... N)` the compiler accepts both but the grammar currently understands only the latter correctly (due to miss interpretation of integral `0` and float `0.`).

I currently do not have a proper process for updating the grammar, my plan was to update on the fly as new ISPC releases are published (if grammar update is required).  The generated EBNF is also very helpful (https://github.com/ispc/ispc/discussions/2529) which I will use to improve the current grammar as I get to it.

Unfortunately there are no universal queries for tree-sitter as different tools/editors have their own flavors.  I currently maintain the queries in https://github.com/nvim-treesitter/nvim-treesitter/tree/master/queries/ispc which work for `nvim`, if somebody wants to use the parser in `emacs` for example they would need to add their own set of queries.  I have added a basic standalone query in https://github.com/fab4100/tree-sitter-ispc/blob/master/queries/highlights.scm which I further plan to use for ISPC syntax highlighting on github (https://github.com/github-linguist/linguist/blob/master/vendor/README.md) because current highlighting is broken (e.g. https://github.com/ispc/ispc/blob/main/stdlib.ispc#L81)

**Reply by dbabokin** [reply]
Thanks for detailed answer.

Is tree-sitter grammar compatible with linguist? I'm working on Textmate grammar for the purpose of introducing it to linguist. It's also is needed for better support through VSCode. But I don't know if automatic conversion from Bison to Textmate is possible and I'm trying to implement it manually.

**Reply by fab4100** [reply]
> Is tree-sitter grammar compatible with linguist?

~Yes~.  Tree-sitter is also used for other features than syntax highlighting on github (e.g. https://github.blog/2021-12-09-introducing-stack-graphs/#creating-stack-graphs-using-tree-sitter).

> I'm working on Textmate grammar for the purpose of introducing it to linguist

I do not want to interfere here, but I think for github functionality a tree-sitter parser for ISPC may be more valuable.  Let me know what you think.

FYI: this PR is currently pending: https://github.com/tree-sitter/tree-sitter/pull/2282

**Reply by fab4100** [reply]
**_Update:_** I wanted to play around with linguist but it seems they are not doing syntax highlighting (which was my initial understanding) and they seem not be converged right now with tree sitter altough they mention some tree-sitter grammars they use.

https://github.com/github-linguist/linguist/discussions/6073

**Comment by pbrubaker** [comment]
Hey @fab4100, it seems we've duplicated some effort here.  My team [(@ethandev](https://github.com/ethandav)) also completed a tree-sitter grammar about two weeks ago! hah! We are going to be using it in our LSP (Language Server Protocol) for better Intellisense like functionality in both vscode and Visual Studio.

We chose to go with a TextMate grammar for syntax highlighting as that's what is used in vscode, and it's supported in Visual Studio.  Those are our main two targets.  But I'm not against using tree-sitter for syntax highlighting.

What are your thoughts on the future maintenance of the ISPC grammar?  Do you want to continue maintaining it and we'll use yours?  Should we merge your work with ours and maintain it as part of the ISPC repo and maintain it from there?

**Reply by fab4100** [reply]
Hi Pete,

> We chose to go with a TextMate grammar for syntax highlighting as that's what is used in vscode, and it's supported in Visual Studio. Those are our main two targets. But I'm not against using tree-sitter for syntax highlighting.

I assume you refer to vscode regarding the TextMate grammar(?)  My main motivation to implement the tree-sitter grammar was mainly syntax highlighting and text objects in `vim` as I am working with ISPC kernles in my applications using this editor.  I am not familiar with vscode and their (future) support for tree-sitter.  Most likely they will support it or based on your writing they likely already do.  For the purpose of this functionality in `vim` (`neovim` precisely) I have added queries that use the https://github.com/fab4100/tree-sitter-ispc parser in the corresponding plugin for the editor already.  The queries included in https://github.com/fab4100/tree-sitter-ispc/tree/master/queries are standalone queries that do not make use of special query syntax that is specific to a tool (`neovim` has some query syntax specific to it, e.g. Lua specific features).  To support syntax highlighting in vscode, either the standalone query mentioned above may be used or specific queries for vscode would have to be written.

> What are your thoughts on the future maintenance of the ISPC grammar? Do you want to continue maintaining it and we'll use yours?

My plan is to maintain the parser as it is already in use for `neovim` (https://github.com/nvim-treesitter/nvim-treesitter/commit/8e817bebf9184cab6c59205e1aed3fdce11c8c51), the name of the grammar rules should not be changed as this will break external queries already in use (instead aliases should be used in the grammar).  The current parser performs as follows on code in the ISPC repository:
```
ISPC tests/*.ispc:
/home/fabs/code/ispc/tests/cfor-gs-double-improve-multidim-1.ispc               0 ms    (MISSING ";" [8, 53] - [8, 53])
/home/fabs/code/ispc/tests/cfor-gs-double-improve-multidim-2.ispc               0 ms    (MISSING ";" [8, 40] - [8, 40])
/home/fabs/code/ispc/tests/cfor-gs-double-improve-multidim-3.ispc               0 ms    (MISSING ";" [8, 40] - [8, 40])
/home/fabs/code/ispc/tests/cfor-gs-double-improve-multidim.ispc                 0 ms    (MISSING ";" [8, 40] - [8, 40])
/home/fabs/code/ispc/tests/cfor-gs-double-improve-multidim-struct-1.ispc        0 ms    (MISSING ";" [14, 40] - [14, 40])
/home/fabs/code/ispc/tests/cfor-gs-double-improve-multidim-struct.ispc          0 ms    (MISSING ";" [12, 40] - [12, 40])
/home/fabs/code/ispc/tests/cfor-gs-improve-multidim-1.ispc                      0 ms    (MISSING ";" [5, 53] - [5, 53])
/home/fabs/code/ispc/tests/cfor-gs-improve-multidim-2.ispc                      0 ms    (MISSING ";" [5, 40] - [5, 40])
/home/fabs/code/ispc/tests/cfor-gs-improve-multidim-3.ispc                      0 ms    (MISSING ";" [5, 40] - [5, 40])
/home/fabs/code/ispc/tests/cfor-gs-improve-multidim.ispc                        0 ms    (MISSING ";" [5, 41] - [5, 41])
/home/fabs/code/ispc/tests/cfor-gs-improve-multidim-struct-1.ispc               0 ms    (MISSING ";" [11, 40] - [11, 40])
/home/fabs/code/ispc/tests/cfor-gs-improve-multidim-struct.ispc                 0 ms    (MISSING ";" [9, 40] - [9, 40])
/home/fabs/code/ispc/tests/cfor-test-89.ispc                                    0 ms    (MISSING ";" [6, 28] - [6, 28])
/home/fabs/code/ispc/tests/cfor-test-91.ispc                                    0 ms    (MISSING ";" [9, 28] - [9, 28])
/home/fabs/code/ispc/tests/cfor-test-96.ispc                                    0 ms    (MISSING ";" [6, 28] - [6, 28])
/home/fabs/code/ispc/tests/intptr.ispc                                          0 ms    (ERROR [6, 4] - [6, 23])
Total parses: 1424; successful parses: 1408; failed parses: 16; success percentage: 98.88%

ISPC tests/lit-tests/*.ispc:
/home/fabs/code/ispc/tests/lit-tests/1673.ispc                                  0 ms    (ERROR [17, 2] - [19, 3])
/home/fabs/code/ispc/tests/lit-tests/1762.ispc                                  0 ms    (ERROR [6, 12] - [7, 30])
/home/fabs/code/ispc/tests/lit-tests/1771.ispc                                  0 ms    (ERROR [23, 4] - [24, 79])
/home/fabs/code/ispc/tests/lit-tests/1788.ispc                                  0 ms    (ERROR [6, 4] - [6, 23])
/home/fabs/code/ispc/tests/lit-tests/861.ispc                                   0 ms    (ERROR [15, 4] - [16, 31])
/home/fabs/code/ispc/tests/lit-tests/ast_stmt.ispc                              0 ms    (ERROR [40, 4] - [42, 19])
/home/fabs/code/ispc/tests/lit-tests/float_literals_errors.ispc                 0 ms    (ERROR [13, 11] - [13, 12])
/home/fabs/code/ispc/tests/lit-tests/fp64_support.ispc                          0 ms    (ERROR [6, 0] - [13, 1])
/home/fabs/code/ispc/tests/lit-tests/func_template_not_supported.ispc           0 ms    (ERROR [6, 21] - [6, 26])
/home/fabs/code/ispc/tests/lit-tests/template_argument_deduction2.ispc          0 ms    (ERROR [12, 31] - [12, 32])
/home/fabs/code/ispc/tests/lit-tests/template_argument_deduction3.ispc          0 ms    (MISSING ";" [14, 38] - [14, 38])
Total parses: 220; successful parses: 209; failed parses: 11; success percentage: 95.00%

ISPC tests_errors/*.ispc:
/home/fabs/code/ispc/tests_errors/func-def-with-typedef.ispc                    0 ms    (MISSING ";" [2, 35] - [2, 35])
/home/fabs/code/ispc/tests_errors/lvalue-2.ispc                                 0 ms    (ERROR [3, 4] - [3, 7])
/home/fabs/code/ispc/tests_errors/lvalue-3.ispc                                 0 ms    (ERROR [4, 4] - [4, 7])
/home/fabs/code/ispc/tests_errors/new-delete-2.ispc                             0 ms    (ERROR [5, 11] - [5, 16])
/home/fabs/code/ispc/tests_errors/old-launch-syntax.ispc                        0 ms    (MISSING identifier [15, 17] - [15, 17])
Total parses: 161; successful parses: 156; failed parses: 5; success percentage: 96.89%

ISPC examples/*:
/home/fabs/code/ispc/examples/cpu/mandelbrot_tasks/mandelbrot_tasks.ispc        0 ms    (ERROR [30, 0] - [82, 1])
/home/fabs/code/ispc/examples/cpu/sort/sort.ispc                                1 ms    (ERROR [114, 94] - [114, 98])
/home/fabs/code/ispc/examples/xpu/simple-dpcpp-l0/simple-dpcpp-l0.ispc          0 ms    (MISSING ";" [32, 35] - [32, 35])
/home/fabs/code/ispc/examples/xpu/callback-esimd/callback-esimd.ispc            0 ms    (MISSING ";" [44, 36] - [44, 36])
/home/fabs/code/ispc/examples/xpu/simple-dpcpp/simple-dpcpp.ispc                0 ms    (MISSING ";" [32, 35] - [32, 35])
/home/fabs/code/ispc/examples/xpu/simple-usm/simple-usm.ispc                    0 ms    (MISSING ";" [31, 35] - [31, 35])
/home/fabs/code/ispc/examples/xpu/usm-mem/usm-mem.ispc                          0 ms    (MISSING ";" [31, 31] - [31, 31])
/home/fabs/code/ispc/examples/xpu/aobench/aobench.ispc                          1 ms    (MISSING ";" [314, 31] - [314, 31])
/home/fabs/code/ispc/examples/xpu/mandelbrot/mandelbrot.ispc                    0 ms    (MISSING ";" [61, 39] - [61, 39])
/home/fabs/code/ispc/examples/xpu/simple-esimd/simple-esimd.ispc                0 ms    (MISSING ";" [37, 35] - [37, 35])
/home/fabs/code/ispc/examples/xpu/simple/simple.ispc                            0 ms    (MISSING ";" [32, 35] - [32, 35])
/home/fabs/code/ispc/examples/xpu/simple-fence/simple-fence.ispc                0 ms    (MISSING ";" [34, 35] - [34, 35])
/home/fabs/code/ispc/examples/xpu/invoke-sycl-aobench/ao.ispc                   0 ms    (MISSING "\n" [3, 18] - [3, 18])
/home/fabs/code/ispc/examples/xpu/noise/noise.ispc                              1 ms    (MISSING ";" [125, 34] - [125, 34])
Total parses: 32; successful parses: 18; failed parses: 14; success percentage: 56.25%
```
Errors are mostly due to intermixed pre-processor statements and `foreach` ranges without spaces between expressions and `...` operator (issues due to number literals interpreted as floats or field declarators if expression is an identifier; implementing this is tricky without lookahead regex support in tree-sitter [will possibly be added in the future]).  The missing semicolons reported are mostly pre-processor macros interpreted as call expressions without `;`. My parser implementation inherits from the C tree-sitter parser and adds the ISPC specific grammar on top.

> Should we merge your work with ours and maintain it as part of the ISPC repo and maintain it from there?

Alternatively, the parser could possibly be moved into the https://github.com/ispc organization if that is preferred.  I think it does not make sense have two parsers, merging would also be possible but the issue I mentioned above could break existing queries in other tools where the parser is already in use. Let me know your thoughts.

**Reply by dbabokin** [reply]
> Alternatively, the parser could possibly be moved into the https://github.com/ispc organization if that is preferred.

That would be my preference. Having the grammar next to the official repo seems to be the most logical. To go forward with it, I think the most important is to decide on the license of the the repo. My preference would be BSD-3C (to match current ISPC license) or Apache-2.0 (as it's the most popular permissive license nowadays, including LLVM, even though with exceptions). Is MIT license was chosen because you inherited from existing C grammar? 

And I think we'll be able to sort out technical questions - like keeping the name of the queries, etc.

And a side note - are you aware of https://github.com/ispc/ispc/blob/main/contrib/ispc.vim ? It's ISPC syntax highlighting for VIM (classic, not NeoVIM).

**Reply by fab4100** [reply]
MIT was chosen because the C parser uses MIT as well.  I do not see a problem with BSD3 or Apache-2.0.
I am aware of `ispc.vim`, was using it before (thanks for it).  I switched to tree-sitter for all my main programming languages due to more accuracy than regex based engines, ISPC was the last missing for which no parser existed.

**Reply by fab4100** [reply]
What are the next steps for migrating the parser repository? (I can change license and push the code to the new repo)

**Reply by dbabokin** [reply]
I've created a private repo and sent you an invite. I'll need to do some paperwork to switch it to public repo, but this should not take too long.

My primary concern is the license in the sense that we need to understand where the prior work comes from, so we are 100% clean.  We'll need to have a "Bill of Material", which explains where the third party code comes from. Either @pbrubaker or myself will need to work with you to build this BOM.

I think you can migrate your repo to this new one and we'll go from there.

**Reply by fab4100** [reply]
I have pushed the code and added a few remarks https://github.com/ispc/ispc.tree-sitter#todo. The Rust crate item is optional and it requires an account on crates.io (GitHub login should work).  The crate can be created via CI but I currently used a crates.io account linked to my GitHub account.

**Reply by MarijnS95** [reply]
> I'll need to do some paperwork to switch it to public repo, but this should not take too long.

Is this still in the works? I'm adding ISPC support to `helix` in https://github.com/helix-editor/helix/pull/8678 and pointing to @fab4100's repository now.

**Reply by pbrubaker** [reply]
Hey @MarijnS95, yes, still in the works. This fell down my todo list but I will follow up and get it done as quickly as possible.

**Comment by aneshlya** [comment]
Apologies for the delayed update—our team has gone through several changes, and we haven’t had the opportunity to revisit this until now. I noticed that a tree-sitter grammar for ISPC was contributed to [tree-sitter-grammars](https://github.com/tree-sitter-grammars), which is great. I think keeping the ISPC grammar in that repository makes sense, so I’d prefer not to move it under the ISPC organization at this point.

@fab4100, are you still maintaining the project? The language continues to evolve with each release, but it looks like the last update to the repository was in 2023. 
I’d love to use [tree-sitter-ispc](https://github.com/tree-sitter-grammars/tree-sitter-ispc) to add ISPC support to Linguist, so we can eventually have ISPC files recognized and highlighted on GitHub.

**Reply by fab4100** [reply]
Hi @aneshlya, my repo has been moved into the tree-sitter-grammars as you mentioned above.  I do have write access to the repo and I track most of the relevant discussions.  I have not used ISPC much lately but I can update it, the grammar inherits from the C grammar and these updates are automatically contained.

**Reply by aneshlya** [reply]
Thanks for your reply! I'll do some testing and get back to you if some changes are needed!

**Reply by aneshlya** [reply]
Linguist still doesn't support tree-sitter grammars so I added a TextMate grammar for ISPC available here: https://github.com/ispc/ispc.syntax and contributed support for Linguist.

**Comment by aneshlya** [comment]
I think we can close this discussion for now.
To sum up ISPC has two grammars now:

- Tree-sitter grammar: [tree-sitter-ispc](https://github.com/tree-sitter-grammars/tree-sitter-ispc)
- TextMate grammar: [ispc.syntax](https://github.com/ispc/ispc.syntax)
